### PR TITLE
Added Spanish

### PR DESCRIPTION
**Pull request by mcliment**
Added Spanish support. I was fun to discover some intrincate language rules on the way.

### CODE REVIEW

**Comment by ploeh** [comment]
Thank you for your interest in contributing to Numsense! It builds and runs all tests without warnings on my machine, so from a technical perspective I think we're good to go :+1:

As I don't read Spanish, I'd like someone who does to review that part of it, if at all possible. I don't expect any errors, but it's always good with a second pair of eyes :smile:

**Comment by ploeh** [comment]
I solicited a review on Twitter: https://twitter.com/ploeh/status/692036372140523521

Please retweet and spread the word :smile:

File: Numsense.UnitTests/SpanishExamples.fs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+[<InlineData(        13, "trece")>]
+[<InlineData(        14, "catorce")>]
+[<InlineData(        15, "quince")>]
+[<InlineData(        16, "dieciséis")>]
+[<InlineData(        17, "diecisiete")>]
+[<InlineData(        18, "dieciocho")>]
+[<InlineData(        19, "diecinueve")>]
+[<InlineData(        20, "veinte")>]
+[<InlineData(        21, "veintiuno")>]
+[<InlineData(        22, "veintidós")>]
```
Comment (FranciscoE-Hudl) [inline Numsense.UnitTests/SpanishExamples.fs]: Accentuation not required.

File: Numsense.UnitTests/SpanishExamples.fs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+[<InlineData(        23, "veintitrés")>]
```
Comment (FranciscoE-Hudl) [inline Numsense.UnitTests/SpanishExamples.fs]: Accentuation not required.

File: Numsense.UnitTests/SpanishExamples.fs [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+[<InlineData(      9000, "nueve mil")>]
+[<InlineData(     10000, "diez mil")>]
+[<InlineData(     12001, "doce mil uno")>]
+[<InlineData(     23456, "veintitrés mil cuatrocientos cincuenta y seis")>]
+[<InlineData(     32109, "treinta y dos mil ciento nueve")>]
+[<InlineData(     40404, "cuarenta mil cuatrocientos cuatro")>]
+[<InlineData(     56789, "cincuenta y seis mil setecientos ochenta y nueve")>]
+[<InlineData(     60015, "sesenta mil quince")>]
+[<InlineData(     71003, "setenta y un mil tres")>]
+[<InlineData(     80522, "ochenta mil quinientos veintidós")>]
```
Comment (FranciscoE-Hudl) [inline Numsense.UnitTests/SpanishExamples.fs]: Accentuation not required.

File: Numsense/Spanish.fs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    | 10 -> "diez"
+    | 11 -> "once"
+    | 12 -> "doce"
+    | 13 -> "trece"
+    | 14 -> "catorce"
+    | 15 -> "quince"
+    | 16 -> "dieciséis"
+    | Between 17 20 x -> simplify "dieci" "" 10 x
+    | 20 -> "veinte"
+    | 22 -> "veintidós"
```
Comment (FranciscoE-Hudl) [inline Numsense/Spanish.fs]: Accentuation not required.

File: Numsense/Spanish.fs [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    | 23 -> "veintitrés"
```
Comment (FranciscoE-Hudl) [inline Numsense/Spanish.fs]: Accentuation not required.

**Comment by mcliment** [comment]
I see other PRs that always add hyphens between each particle even if that's not correct in their languages (i.e. in Spanish they are never used), what's the point on that?

**Comment by ncave** [comment]
I have the same question, is there a reason not to switch to space delimiters for the generation? Parsing can still handle both space and hyphen.

**Comment by ploeh** [comment]
We had [a similar discussion about the German implementation](https://github.com/ploeh/Numsense/pull/13#issuecomment-170470365). Essentially, if at all 'legal' in the language, it'd be useful for any consumer of the Numsense API to have a way to further manipulate the generated string, if necessary.

Leaving in *some* delimiter is a low-tech way of enabling that.

As an example, given the string "forty-two", a client can always remove the hyphen if it prefers "fortytwo". It also has the option to convert the numeral to "FortyTwo", or do all sorts of weird and unpredictable things. On the other hand, if we gave the client the string "fortytwo", there would be no easy way to generate "forty-two" from that string.

The hyphen was chosen for Danish because it's an optional, but correct, way to spell Danish numerals. As far as I can tell, the same is true for English; case in point, [Nineteen Eighty - Four](http://amzn.to/1nosBil) (although a bit less consistently applied here).

If Spanish doesn't allow the use of a hyphen, then so be it. If there's another delimiter we can use, it'd still be preferable, so that we don't leave clients without options.

If it's absolutely never OK to spell Spanish numerals with delimiters, then perhaps we should consider also exposing an alternative function that returns the numeral as a list of small chunks, instead of as an opaque string.

Now that I think about it, perhaps we should do that for all the languages...

**Comment by mcliment** [comment]
Oh, I see.

In Spanish is not correct to use a hyphen but can be done if it's for technical reasons. On the string to number conversion it's not a problem to replace all spaces by hyphens to be consistent with other implementations.
But for number to string I see that if I use hyphens instead of spaces, the output is not grammatically correct. Replacing them by spaces by the consumer can do but seems weird to me, as he will need to know what is correct Spanish.
And there are no _alternative_ ways to write the numbers, every non single word is separated by a space.

On the other hand, I'm not sure on what to do with diacritics and accents. Should they be removed when parsing the strings? Strictly speaking it's not correct language but can slightly broaden the scope of inputs.

One more thing is gender. My PR only deals with masculine cardinals but they are a bit different when used with feminine nouns: "veintiún coches" vs "veintiuna palabras" (twenty-one cars vs twenty-one words). This might be an enhancement for the future because other languages may change too. 

And finally, just say that I'm almost sure that this parser is valid for every Spanish variant as numbers don't change.

**Comment by ploeh** [comment]
> And there are no alternative ways to write the numbers

OK. I anticipated that this would happen for some language, sooner or later.

How about adding a function called something like `toSpanishElements`? I'm open to suggestions regarding the name, but the type would be `int -> string list`. The implementation of `toSpanish` would essentially be a call to `toSpanishElements`.

This design could be back-ported to the other languages (but not in this pull request, and not necessarily by you).

Does that sound like a way forward?

> On the other hand, I'm not sure on what to do with diacritics and accents. Should they be removed when parsing the strings? Strictly speaking it's not correct language but can slightly broaden the scope of inputs.

Numsense follows [Postel's law](https://en.wikipedia.org/wiki/Robustness_principle), so a parsing function can be a [Tolerant Reader](http://martinfowler.com/bliki/TolerantReader.html) and accept incorrect input, as long as it's unambiguous. If the current pull request doesn't already do this, though, I think this could be a topic for a future request.

>  My PR only deals with masculine cardinals but they are a bit different when used with feminine nouns: "veintiún coches" vs "veintiuna palabras" (twenty-one cars vs twenty-one words). This might be an enhancement for the future because other languages may change too.

I'm fine with adding this in the future :+1:

**Comment by mcliment** [comment]
> Does that sound like a way forward?

Seems good to have a function to get elements for machine use and another for human grammar perfect readable strings.

Then I have one more doubt -sorry-. In Spanish you have _tres mil_ which is _three thousand_ and _trescientos_ which is _three hundred_. So, _-cientos_ is actually _hundreds_ but can't be broken in two words. Do you think this would be two different elements or just one? From the language perspective is a single word.

> If the current pull request doesn't already do this, though, I think this could be a topic for a future request.

It does not and I'll leave the diacritics thing for an improvement.

**Comment by ploeh** [comment]
> From the language perspective is a single word.

If you say it's a single word, then it's a single word :smile:

**Comment by mcliment** [comment]
Fixed conflicts with master

**Comment by ploeh** [comment]
Sorry, but @SabotageAndi beat you by twenty minutes, so I've now pulled #35, rendering this pull request in conflict yet again.

I'm beginning to realise that I've organised the code in this repository in such a way that it generates conflicts because all new languages have to touch the same files. That's my mistake - sorry about that :flushed: 

Could you, though, address the conflicts?

**Comment by mcliment** [comment]
Done, this merge was easier as the new languages did not conflict.

**Comment by ploeh** [comment]
Thank you for your contribution! It's now live as [Numsense 0.11.0](https://www.nuget.org/packages/Numsense/0.11.0).

**Pull request merged**